# `OperationResult`

A struct to represent a operation result with a success or failure state.

The `OperationResult` is similar to the `ValidableResult` but state (success or failure) can not be changed after creation.

The following demonstrates how to use the `OperationResult` struct.

## Creating a success result

```csharp
var result = new OperationResult;

// or

OperationResult result = new();
```

## Creating a failure result

The sample below shows how to create a failure result:

```csharp
var errors = new ResultErrors();
errors += ResultMessage.Error("Error message");

var result = new OperationResult(errors);
```

In the example above, the `OperationResult` is created via the constructor, passing `ResultErrors` as a parameter.

You can create the `OperationResult` using the implicit conversion from `ResultErrors` to `OperationResult`:

```csharp
var errors = new ResultErrors();
errors += ResultMessage.Error("Error message");

OperationResult result = errors;
```

There is also an implicit conversion from the error message:

```csharp
OperationResult result = ResultMessage.Error("Error message");
```

Once a successful `OperationResult` has been created, it is not possible to add errors to it.
If you try to add errors to a successful `OperationResult`, an exception will be thrown.

Here's an example that throws an exception when trying to add errors to a successful `OperationResult`:

```csharp
var result = new OperationResult();

result += ResultMessage.Error("Error message"); // throws InvalidOperationException
```

In the case of a failed `OperationResult`, you can add errors to it:

```csharp
// create a failure result
OperationResult result = ResultMessage.Error("Error message 1");

// add errors
result += ResultMessage.Error("Error message 2");
result += ResultMessage.Error("Error message 3");
```

## Working with the `OperationResult`

The `OperationResult` has the same methods as the `ValidableResult`,
and some additional methods to work with the success or failure state.

The following methods are available:
- `TryGetError`
- `Convert`
- `Match`
- `TryConvertError<TOtherError>`
- `IsSuccessOrGetError`
- `IsSuccessOrConvertError`

The first three were presented in the `ValidableResult` section, now let's look at the other methods.

### `TryConvertError<TOtherError>`

This method is similar to `TryGetError`, returning a `bool`, being true when the result is a failure.
However, instead of the output being a `ResultError`, it will be the result of converting the `ResultError` into a `TOtherError`.

The conversion will only take place if the result is a failure.

Here's an example of how to use it:

